fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java

```diff
@@ -0,0 +1,338 @@
+        this.exist = exist;
+    }
+
+    public List<List<String>> getClustersInfo() {
+        return procNode.fetchResult().getRows();
+    }
+
+
+    public void addCluster() throws DdlException {
+        writeLock();
+        try {
+            if (clusters.size() < maxCluster) {
+                long workerGroupId = GlobalStateMgr.getCurrentStarOSAgent().createWorkerGroup();
+                long clusterId = GlobalStateMgr.getCurrentState().getNextId();
+                Cluster cluster = new Cluster(clusterId, workerGroupId);
+                clusters.put(clusterId, cluster);
+                OpClusterLog opClusterLog = new OpClusterLog(this.getFullName(), cluster);
+                GlobalStateMgr.getCurrentState().getEditLog().logAddCluster(opClusterLog);
+
+                state = WarehouseState.RUNNING;
+
+                // for debug
+                LOG.info("add cluster {} for warehouse {} ", clusterId, name);
+            } else {
+                throw new DdlException("cluster num of " + name + " exceed max cluster limit " + maxCluster);
+            }
+        } finally {
+            writeUnLock();
+        }
+    }
+
+    public void removeCluster() throws DdlException {
+        writeLock();
+        try {
+            if (clusters.size() > minCluster) {
+                Long clusterId = clusters.keySet().stream().findFirst().get();
+                Cluster cluster = clusters.get(clusterId);
+                GlobalStateMgr.getCurrentStarOSAgent().deleteWorkerGroup(cluster.getWorkerGroupId());
+                clusters.remove(clusterId);
+                OpClusterLog opClusterLog = new OpClusterLog(this.getFullName(), cluster);
+                GlobalStateMgr.getCurrentState().getEditLog().logRemoveCluster(opClusterLog);
+
+                // for debug
+                LOG.info("remove cluster {} for warehouse {} ", clusterId, name);
+            } else {
+                throw new DdlException("cluster num of " + name + " exceed min cluster limit " + minCluster);
+            }
+        } finally {
+            writeUnLock();
+        }
+    }
+
+    public void replayAddCluster(OpClusterLog log) {
+        writeLock();
+        try {
+            Cluster cluster = log.getCluster();
+            clusters.put(cluster.getId(), cluster);
+            this.state = WarehouseState.RUNNING;
+            // for debug
+            LOG.info("replay add cluster {} for warehouse {} ", cluster.getId(), name);
+        } finally {
+            writeUnLock();
+        }
+    }
+
+    public void replayRemoveCluster(OpClusterLog log)  {
+        writeLock();
+        try {
+            Cluster cluster = log.getCluster();
+            clusters.remove(cluster.getId());
+            // for debug
+            LOG.info("replay remove cluster {} for warehouse {} ", cluster.getId(), name);
+        } finally {
+            writeUnLock();
+        }
+    }
+
+    public void suspend() {
+        state = WarehouseState.SUSPENDED;
```

Comment:
just a remind, we also need to persist this state and release all the cluster resource here